### PR TITLE
Update readme with consumers guide and features

### PR DESCRIPTION
**Pull request by XiongKezhi**
- [x] Update readme

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -3,12 +3,72 @@
 [![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](https://issues.jenkins-ci.org/issues/?jql=component%20%3D%20checks-api-plugin)
 [![Jenkins](https://ci.jenkins.io/job/Plugins/job/checks-api-plugin/job/master/badge/icon?subject=Jenkins%20CI)](https://ci.jenkins.io/job/Plugins/job/checks-api-plugin/job/master/)
 [![GitHub Actions](https://github.com/jenkinsci/checks-api-plugin/workflows/CI/badge.svg?branch=master)](https://github.com/jenkinsci/checks-api-plugin/actions)
+[![codecov](https://codecov.io/gh/jenkinsci/checks-api-plugin/branch/master/graph/badge.svg)](https://codecov.io/gh/jenkinsci/checks-api-plugin)
 
+Inspired by the [GitHub Checks API](https://docs.github.com/en/rest/reference/checks#runs), this plugin aims to provide a general API to allow Jenkins plugins publishing checks (or reports) to remote source code management (SCM) platforms (e.g. GitHub, GitLab, BitBucket, etc.).
+By consuming this API, other plugins can publish check with customized parameters for a Jenkins build, such as status, summary, warnings, code annotations, or even images.
```
Comment (timja) [inline README.md:9]: Let's add some line breaks to break up the paragraphs this is a new topic.

```suggestion

By consuming this API, other plugins can publish check with customized parameters for a Jenkins build, such as status, summary, warnings, code annotations, or even images.
```

File: README.md [hunk 2]
```diff
@@ -3,12 +3,72 @@
...
+Then, the implementations of this API will decide on how to make use of these parameters and where to publish the checks.
```
Comment (timja) [inline README.md:10]:
```suggestion
Implementations of this API decide on how to make use of these parameters and where to publish the checks.
```

File: README.md [hunk 3]
```diff
@@ -3,12 +3,72 @@
...
 
-This a [Google Summer of Code](https://summerofcode.withgoogle.com/) project for [Jenkins Organization](https://www.jenkins.io/), it is still under development.
+Current consumers of this plugin include [Warnings Next Generation Plugin](https://github.com/jenkinsci/warnings-ng-plugin) and [Code Coverage API Plugin](https://github.com/jenkinsci/code-coverage-api-plugin); current implementations include [GitHub Checks Plugin](https://github.com/jenkinsci/github-checks-plugin).
```
Comment (timja) [inline README.md:12]: I think this looks better, might be nice to add integration pull requests so others can see how to add their own one

```suggestion
Known consumers:
* [Warnings Next Generation Plugin](https://plugins.jenkins.io/warnings-ng)
* [Code Coverage API Plugin](https://plugins.jenkins.io/code-coverage-api)

Implementations:
* [GitHub Checks Plugin](https://plugins.jenkins.io/github-checks)
```

File: README.md [hunk 4]
```diff
@@ -3,12 +3,72 @@
...
 
-## Useful links:
-* [Project Idea](https://www.jenkins.io/projects/gsoc/2020/project-ideas/github-checks/)
-* [Project Design](https://docs.google.com/document/d/1hVQTd9jKw0sx8JQR8KjbM-7lWW84e2vFjmmkzBpBbSk/edit?usp=sharing)
-* [Meeting Notes](https://docs.google.com/document/d/1TZLmu3nBPbwUjzLVYGnV_YtYvmzxzw6A4eEVYpbmi3Y/edit?usp=sharing)
+## Consumers Guide
```
Comment (timja) [inline README.md:14]: I think the guide belongs in its own file in the docs directory.

I would consider moving pipeline above this, as users will want this.

File: README.md [hunk 5]
```diff
@@ -3,12 +3,72 @@
...
+ChecksDetails details = new ChecksDetailsBuilder()
+ .withName("Jenkins CI")
+ .withStatus(ChecksStatus.COMPLETED)
+ .withConclusion(ChecksConclusion.SUCCESS)
+ .withDetailsURL("https://ci.jenkins.io")
+ .withCompletedAt(LocalDateTime.now(ZoneOffset.UTC))
+ .build();
+```
+
+Then you can create a publisher based on a Jenkins `run` to publish the check you just constructed:
```
Comment (timja) [inline README.md:31]:
```suggestion
Then you can create a publisher based on a Jenkins `Run` to publish the check you just constructed:
```

File: README.md [hunk 6]
```diff
@@ -3,12 +3,72 @@
...
+
+Imagine that you want to publish a check with some details of a Jenkins build. 
+You need to first construct a `ChecksDetails` object:
+
+```
+ChecksDetails details = new ChecksDetailsBuilder()
+ .withName("Jenkins CI")
+ .withStatus(ChecksStatus.COMPLETED)
+ .withConclusion(ChecksConclusion.SUCCESS)
+ .withDetailsURL("https://ci.jenkins.io")
```
Comment (timja) [inline README.md:26]: probably should include in the example how to get the url i.e. from the display url api?

File: README.md [hunk 7]
```diff
@@ -3,12 +3,72 @@
...
+
+### Checks Parameters
+
+The checks are highly customized by consumers due to a number of optional parameters provided.
+Consumers can set these parameters through the checks models:
+
+- `ChecksDetails`: the top-level model of a check, including all other models and some basic parameters like status and conclusion;
+- `ChecksOutput`: the output of a check, providing some displayed details like title, summary, URL, and code annotations;
+- `ChecksAnnotation`: a code annotation of a check, providing a specific comment for one or multiple lines of code;
+- `ChecksImage`: an image of a check, providing an intuitive graph for the build like issues trend, coverage chart;
```
Comment (timja) [inline README.md:48]:
```suggestion
- `ChecksImage`: an image of a check, providing an intuitive graph for the build like issues trend, coverage chart, must be a public URL that GitHub can fetch from
```

File: README.md [hunk 8]
```diff
@@ -3,12 +3,72 @@
...
+
+The publisher returned is based on the implementations you installed on your Jenkins instance.
+
+### Checks Parameters
+
+The checks are highly customized by consumers due to a number of optional parameters provided.
+Consumers can set these parameters through the checks models:
+
+- `ChecksDetails`: the top-level model of a check, including all other models and some basic parameters like status and conclusion;
+- `ChecksOutput`: the output of a check, providing some displayed details like title, summary, URL, and code annotations;
```
Comment (timja) [inline README.md:46]:
```suggestion
- `ChecksOutput`: the output of a check, providing some displayed details like title, summary, URL, and code annotations
```

File: README.md [hunk 9]
```diff
@@ -3,12 +3,72 @@
...
+```
+
+The publisher returned is based on the implementations you installed on your Jenkins instance.
+
+### Checks Parameters
+
+The checks are highly customized by consumers due to a number of optional parameters provided.
+Consumers can set these parameters through the checks models:
+
+- `ChecksDetails`: the top-level model of a check, including all other models and some basic parameters like status and conclusion;
```
Comment (timja) [inline README.md:45]:
```suggestion
- `ChecksDetails`: the top-level model of a check, including all other models and some basic parameters like status and conclusion
```

File: README.md [hunk 10]
```diff
@@ -3,12 +3,72 @@
...
+- `ChecksOutput`: the output of a check, providing some displayed details like title, summary, URL, and code annotations;
+- `ChecksAnnotation`: a code annotation of a check, providing a specific comment for one or multiple lines of code;
+- `ChecksImage`: an image of a check, providing an intuitive graph for the build like issues trend, coverage chart;
+- `ChecksAction`: an action of a check, providing further actions to be performed by users, like rerun a Jenkins build;
```
Comment (timja) [inline README.md:49]:
```suggestion
- `ChecksAction`: an action of a check, providing further actions to be performed by users, like rerun a Jenkins build
```

File: README.md [hunk 11]
```diff
@@ -3,12 +3,72 @@
...
+### Checks Publishers
+
+The publishers are created through the static factory method (`fromRun` or `fromJob`) of `ChecksPublisherFactory`.
+The factory will iterate all available implementations of the `ChecksPublisher` in order to find the suitable publisher for the Jenkins `Run` or `Job`.
+
+## Embedded Features
+
+### Build Status Check
+
+![GitHub Status](docs/images/github-status.png)
```
Comment (timja) [inline README.md:60]: should also include a screenshot showing the checks on ci.jenkins.io, probably near the top of the README

**Review by timja — APPROVED** [review]
no blockers but a number of suggested improvements.

File: README.md (hunk 11 above)
Comment (XiongKezhi) [reply README.md:60]: I prefer to add that screenshot in the GitHub implementation, but I think I should add a section (and a link to the actions tab of this repo) after the known implementations to tell them where to take a look at the results.

File: README.md [hunk 12]
```diff
@@ -3,12 +3,43 @@
...
-* [Project Design](https://docs.google.com/document/d/1hVQTd9jKw0sx8JQR8KjbM-7lWW84e2vFjmmkzBpBbSk/edit?usp=sharing)
-* [Meeting Notes](https://docs.google.com/document/d/1TZLmu3nBPbwUjzLVYGnV_YtYvmzxzw6A4eEVYpbmi3Y/edit?usp=sharing)
+Known consumers:
+* [Warnings Next Generation Plugin](https://plugins.jenkins.io/warnings-ng)
+* [Code Coverage API Plugin](https://plugins.jenkins.io/code-coverage-api)
 
+Implementations:
+* [GitHub Checks Plugin](https://plugins.jenkins.io/github-checks)
+
+This plugin, alone with it's [GitHub implementation](https://plugins.jenkins.io/github-checks), has been installed on [ci.jenkins.io](https://ci.jenkins.io/Plugins) to help maintain hundreds of Jenkins plugins. You can take a look at the [action](https://github.com/jenkinsci/checks-api-plugin/runs/1025532156) for this repository or other plugin repositories under [Jenkins organization](https://github.com/jenkinsci) for the results.
```
Comment (timja) [inline README.md:20]:
```suggestion
This plugin, along with it's [GitHub implementation](https://plugins.jenkins.io/github-checks), has been installed on [ci.jenkins.io](https://ci.jenkins.io/Plugins) to help maintain over 1500 Jenkins plugins. You can take a look at the [action](https://github.com/jenkinsci/checks-api-plugin/runs/1025532156) for this repository or other plugin repositories under [Jenkins organization](https://github.com/jenkinsci) for the results.
```

File: docs/consumers-guide.md [hunk 13]
```diff
@@ -0,0 +1,42 @@
+
+## Consumers Guide
+
+### A Simple Example
+
+Imagine that you want to publish a check with some details of a Jenkins build.
```
Comment (timja) [inline docs/consumers-guide.md:6]:
```suggestion
Imagine that you want to publish a check with some details of a Jenkins build. 

```

File: docs/consumers-guide.md [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+ChecksPublisher publisher = ChecksPublisher.fromRun(run);
+publisher.publish(details);
+```
+
+The publisher returned is based on the implementations you installed on your Jenkins instance.
+
+### Checks Parameters
+
+The checks are highly customized by consumers due to a number of optional parameters provided.
+Consumers can set these parameters through the checks models:
```
Comment (timja) [inline docs/consumers-guide.md:31]:
```suggestion

Consumers can set these parameters through the checks models:
```

File: README.md [hunk 15]
```diff
@@ -3,12 +3,43 @@
...
+Instead of depending on consumers plugins, the users can publish their checks directly in the pipeline script:
+
+```
+publishChecks name: 'example', title: 'Pipeline Check', summary: 'check through pipeline', text: 'you can publish checks in pipeline script', detailsURL: 'https://github.com/jenkinsci/checks-api-plugin#pipeline-usage'
+```
+
+## Guides
+
+- [Consumers Guide](docs/consumers-guide.md)
+- [Implementation Guide](docs/implementation-guide.md)
```
Comment (timja) [inline README.md:41]: doesn't appear to exist?

it should be added in the PR that adds it if I'm not misunderstanding

```suggestion
```

File: README.md [hunk 16]
```diff
@@ -3,12 +3,43 @@
...
+
+## Acknowledgements
+
+This plugin is started as a [Google Summer of Code 2020 project](https://summerofcode.withgoogle.com/projects/#5139745388101632), you can find more about it on [Jenkins GSoC SIG](https://www.jenkins.io/sigs/gsoc/).
```
Comment (timja) [inline README.md:45]:
```suggestion
This plugin was started as a [Google Summer of Code 2020 project](https://summerofcode.withgoogle.com/projects/#5139745388101632), you can find more about it on [Jenkins GSoC SIG](https://www.jenkins.io/sigs/gsoc/).
```

File: README.md [hunk 17]
```diff
@@ -3,12 +3,43 @@
...
+```
+publishChecks name: 'example', title: 'Pipeline Check', summary: 'check through pipeline', text: 'you can publish checks in pipeline script', detailsURL: 'https://github.com/jenkinsci/checks-api-plugin#pipeline-usage'
+```
+
+## Guides
+
+- [Consumers Guide](docs/consumers-guide.md)
+- [Implementation Guide](docs/implementation-guide.md)
+
+## Acknowledgements
```
Comment (timja) [inline README.md:43]: probably worth adding the default sections from the archetype README,

Contributing and License?

e.g.
https://github.com/timja/jenkins-junit-postgresql-plugin#contributing

File: README.md [hunk 18]
```diff
@@ -17,7 +17,7 @@ Known consumers:
 Implementations:
 * [GitHub Checks Plugin](https://plugins.jenkins.io/github-checks)
 
-This plugin, alone with it's [GitHub implementation](https://plugins.jenkins.io/github-checks), has been installed on [ci.jenkins.io](https://ci.jenkins.io/Plugins) to help maintain hundreds of Jenkins plugins. You can take a look at the [action](https://github.com/jenkinsci/checks-api-plugin/runs/1025532156) for this repository or other plugin repositories under [Jenkins organization](https://github.com/jenkinsci) for the results.
+This plugin, along with it's [GitHub implementation](https://plugins.jenkins.io/github-checks), has been installed on [ci.jenkins.io](https://ci.jenkins.io/Plugins) to help maintain over 1500 of Jenkins plugins. You can take a look at the [action](https://github.com/jenkinsci/checks-api-plugin/runs/1025532156) for this repository or other plugin repositories under [Jenkins organization](https://github.com/jenkinsci) for the results.
```
Comment (timja) [inline README.md:20]:
```suggestion
This plugin, along with it's [GitHub implementation](https://plugins.jenkins.io/github-checks), has been installed on [ci.jenkins.io](https://ci.jenkins.io/Plugins) to help maintain over 1500 Jenkins plugins. You can take a look at the [action](https://github.com/jenkinsci/checks-api-plugin/runs/1025532156) for this repository or other plugin repositories under [Jenkins organization](https://github.com/jenkinsci) for the results.
```

**Review by timja — APPROVED** [review]

**Pull request merged**